### ISSUE TITLE
Microsoft.OData.Client 7.0

### ISSUE THREAD

**Issue by meancrazy** [OData/odata.net#731]
Please update Microsoft.OData.Client nuget package to version 7.0.

**Comment by patrickmichalina** [comment]
Thanks - not sure why they haven't shipped it yet...

**Comment by MAOliver** [comment]
I have an application using the 6.15 version of the client to contact a service which is using 7.0 with a ComplexType w/ a NavigationProperty.

The client currently fails to instantiate. I can make it work by creating an intermediate complextype and mapping my entity to the complextype and changing my complextype to only reference other complextypes. Is there another, more elegant way to work around this situation while I wait for the new client?

**Comment by earloc** [comment]
any near plans for shipping this?

**Comment by AlanWong-MS** [comment]
Hi everyone, Microsoft.OData.Client 7.x is on our radar (it has not been released with the recent ODL 7.2.0). We're currently sorting out any possible breaking changes with the library and will proceed to publish it when we're ready.

Thanks for your patience!

**Comment by Slimicus** [comment]
@MAOliver do you have any more documentation on your workaround for this issue? I'm currently facing the same error, the edmx won't parse because of a NavigationProperty in the ComplexType. Thanks!

**Comment by MAOliver** [comment]
@Slimicus My situation arose from trying to convert a pre-existing service to OData. I have an entry point that was originally a WebAPI entrypoint whose data contract was created as "XSD-first" and there are several container objects that don't actually map to an entity, but they contain entities themselves.

To solve the immediate issue, I just removed navigation properties from the complex types and replaced them with a complex type that I populate from the entity that the navigation property was pointing to. It means that anyone calling the endpoint won't be able to traverse to the entity endpoint that populated that complextype, but I didn't want to lose the ability to generate clients.

It's hard for me to be clearer about what I did without creating a code sample to show you. Is this clear enough?

**Comment by wmalgadey** [comment]
any update on this? nearly 2 month later?

**Comment by JamieG** [comment]
I'm researching tooling for a series of upcoming projects and I really wanted to use OData, I'm extremely disappointed with the state of the tooling and the lack of VS2017 support. I've attempted to workaround the issues by forking the repo but the solution is a cross referencing headache and the client template is a nightmare, is a 5000 line ttinclude really the best was to go? The whole thing needs a reboot and a rewrite for .NET Core. Is this being neglected because the core team are going in a different direction, should we be looking at an alternative solution?

**Comment by MAOliver** [comment]
@JamieG For data services, I don't think there is another option like OData. GraphQL is promising, but I strongly dislike that you POST for everything. The ASP.Net OData library also has a ODataSwaggerConverter which will translate your OData model into a Swagger/Open API doc that you can expose to allow an integrator to use Open API as the specification to generate their client.

**Comment by patrickmichalina** [comment]
Looks like OData only has a small graveyard team left. Too bad, looked promising.

**Comment by robertmclaws** [comment]
@patrickmichalina Actually, quite the opposite. I get e-mails for every checkin and comment, and I've gotten like 500 so far this month. They are bigger and better than ever, just SUPER busy refactoring the code and getting it ready for both ASP.NET 5 and ASP.NET Core.

Also, they've actually already released v7.0 and are onto 7.3 beta now.

**Comment by patrickmichalina** [comment]
@robertmclaws that is good news!

**Comment by JamieG** [comment]
I hope so, like I said I think OData would be a great tool what I'm scoping, looking forward to getting it up and going in ASP.NET Core!

**Comment by robertmclaws** [comment]
@AlanWong-MS Before you go to ship 7.0, can we chat briefly about how the T4 templates work? I think there would be value in breaking out the code generation similar to how the Entity Framework EDMX code generation separates out the Context generation from the Objects generation. 

The reason being is that I think the objects that are generated (implementing `INotifyPropertyChanging` and `INotifyPropertyChanged`) would be really useful as core observables that can be data-bound in WPF and UWP UIs. If done right, I think it could generate a single assembly of objects that are useful both on the client AND the server.

I've been stewing on this idea for a while now, and I think there's a chance to implement it now and let people do some really cool things.

Thanks!

**Comment by AlanWong-MS** [comment]
@robertmclaws Can you shoot this one to odatafeedback@microsoft.com? I'd like to include others on this input.

**Comment by NetTecture** [comment]
@AlanWong-MS Is there any timeframe for an update on the client libraries? Like many others I am stuck without VS support - and in my case without support for server and client on the same version (as I actually maintain a service that exposes OData but also uses OData to connect to other services). The lack of tooling and the versioning nightmare that was produced due to not publishing updates has really had negative consequences on work planning work there.

**Comment by AlanWong-MS** [comment]
@NetTecture I'm targeting end of September / first week of October. For OData Client to be released, we need tests for the NET Standard version of the binary in 7.x and for such coverage, I narrowed down two test projects that I'm porting to NET Core. I just finished one of the projects last week and I'm working on the second one; it's taking a while due to the various project dependencies for the test project and the non-1-to-1 API translation from NetFx to NetCore. Trying to work on it as fast as I can without losing quality--I want this out the door as much as everyone does. :)

**Comment by AlanWong-MS** [comment]
Just wanted to give a quick update on my progress: I am still working on the test coverage for our Net Standard OData Client library. (Had a few gaps of being out of office and working on other items.) I recently got past what I believe was the biggest hurdle in all this, which was compilation. Additionally, for this release, we decided that we'll remove some deprecated code from OData Client. With diligence, I'm aiming to have the new library released next week--will keep this thread posted.

**Comment by NetTecture** [comment]
@AlanWong-MS  Very good, please do so. I would love to be able to get my hands on something when VS 15.4 is released, which I assume will happen mid monthish (wih the new windows release, which IIRC also has .NET 4.7.1).

**Comment by tchuan** [comment]
@AlanWong-MS , Greate news. We are replacing xrm tooling lib with OData Api to access Dynamics Crm, encountered some problems but resolved. Look forward to this new release.

**Comment by AlanWong-MS** [comment]
Alright, I came across some road bumps on failed test cases but resolved them this week, and I've submitted a PR #971 with test coverage on the new OData Client library. We're still discussing internally for items left for the release, including the aforementioned removal of some deprecated code. We're considering a beta release to get the library exposed sooner pending how much work is left. I'll be out of office for a week and a half--will keep everyone posted when I return!

**Comment by NetTecture** [comment]
Get it out, please. Make a beta, preview, alpha, whatever. A lot of us are stuck at the moment - totally. I would expect that I have about a month of work before I can issue an alpha or beta on my end of the new integration - but the earlier I can start the better.

**Comment by AlanWong-MS** [comment]
Alright, few updates:
- We have a green light to release beta; this will happen sometime on Monday, October 30 PST.
- To promote out of beta, we have to make [OData Connected Service](https://marketplace.visualstudio.com/items?itemName=laylaliu.ODataConnectedService) ([GitHub Issue 54](https://github.com/OData/lab/issues/54)) and [OData v4 Client Code Generator](https://marketplace.visualstudio.com/items?itemName=bingl.ODatav4ClientCodeGenerator) compatible with the new OData Client 7.x. I will be working on this next and I don't anticipate it coming across as many problems as setting up tests for OData Client.
- We will be removing Net45 and PortableLib binaries from the Nuget package of OData Client 7.x; the package will contain only NetStandard 1.1. Please provide feedback on whether the removed libraries are needed for your scenario and if they are needed, how you are using them.

Thank you all for the continued interest and patience!

**Comment by NetTecture** [comment]
What about NetStandard 2.0? Because I do not think anyone cares about 1.1 at this point.

**Comment by AlanWong-MS** [comment]
@NetTecture, I do hear you and I empathize with you and everyone else who would love to see the library go to NetStandard 2.0--development has been so much easier for me on 2.0. At this time, we're trying to make the change very low impact to end user projects by targeting the minimum NetStandard version that correlated with the PCL that we released. Per the chart at the bottom, it's [PCL 111 <-> NetStandard 1.1](https://docs.microsoft.com/en-us/dotnet/standard/net-standard).

We do know of users who aren't able to go past 1.1; I'd encourage you to file a new issue as a community-based initiative for us to drive a move towards 2.0.

**Comment by tchuan** [comment]
It seems you have postponed the beta release?

**Comment by AlanWong-MS** [comment]
@tchuan, I checked the status and it turns out that we're making changes in another set of binaries that we're releasing in the beta. As far as I know, OData Client itself is still on green light. The beta package should come out within this week.

**Comment by grahamehorner** [comment]
@AlanWong-MS any update on odata client ?

**Comment by AlanWong-MS** [comment]
@grahamehorner, we have the beta package released for [OData Client 7.4.0](https://www.nuget.org/packages/Microsoft.OData.Client/). We're in the process of gathering feedback before we move the package out of beta.

**Comment by robertmclaws** [comment]
@AlanWong-MS Any word on the status of the code generation system that supports Client 7.4.0?

**Comment by AlanWong-MS** [comment]
@robertmclaws, I recently wrapped up the migration of the OData Client Code Gen extension to both VS2017 (apparently for the one that's currently in the VSIX gallery, a dev can install the extension onto VS2017 but it won't function correctly) and OData Client 7.4.0. There's a [bug that impacts user experience that I filed for VS2017](https://developercommunity.visualstudio.com/content/problem/143332/vsix-extension-doesnt-retrieve-required-assemblies.html); as far as I know, I addressed the compatibility issues with VS2017 on OData end so the team is discussing whether we should wait for that aforementioned bug to be fixed before we release the new VSIX. This week, I'm on a rotation so I won't be able to release the extension, but next week should be feasible.

TL;DR of the bug I linked: once you install the extension and utilize it to generate your proxy code, Visual Studio is supposed to pull in the necessary references into the project for you. It doesn't for VS2017, whereas it did for VS2015. The workaround is to manually add the references yourself (VS2017 will provide info on which ones are missing), and then the project will build fine.

I appreciate any feedback on whether we should go ahead and release the extension with the bug.

**Comment by robertmclaws** [comment]
@AlanWong-MS The link I mentioned in my last post is a place where you can post text VSIX packages before they go into production. You can do it by [using AppVeyor via the Extensibility Tools](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.ExtensibilityTools), or by using [this powershell script in your build](http://vsixgallery.com/guide/dev/).

If you put it up there so I can try it, I can let you know whether I think it can be published in its current state.

Thanks!

**Comment by AlanWong-MS** [comment]
@robertmclaws, thanks for the resources! I pushed a beta package to the [VSIX Gallery server](http://vsixgallery.com/extension/ODataT4ItemTemplate.7a8976ef-0030-47ec-bd10-e8cdf7a2ccb2/). A few words to be explicit: the package itself is not signed as it's not an official beta package and the server is not an official feed supported by Microsoft.

The few missing dependencies that I had to manually add after I used the code generator are:
1. System
2. System.Data
3. System.Data.DataSetExtensions
4. System.Net.Http
5. System.XML
6. System.Xml.Linq

Thanks again for any feedback!

**Comment by mprigge** [comment]
@AlanWong-MS I've been sort of forced into testing the 7.4.0-beta client code because I am in a situation where I need to support both the server and client in the same project. I'm finding that linq calls I had previously used with the 6.x client fail with a NotSupportedException reading:

> This target framework does not enable you to directly enumerate over a data service query. This is because enumeration automatically sends a synchronous request to the data service. Because this framework only supports asynchronous operations, you must instead call the BeginExecute and EndExecute methods to obtain a query result that supports enumeration.

I suspect this may be due to some difference in the removed net45 library and the PCL 111/NetStandard 1.1 library. This is the only reason I mention this here versus opening a new issue (because you had asked above whether there were issues introduced by removing the other libraries).

An example use case is this fairly simple linq query:

```
var person = api.People
    .Expand(x => x.Market)
    .Where(x => x.ActiveDirectoryRefs.Any(y => y.Sid == sid && y.Domain == domain))
    .SingleOrDefault();
```
Previously, that'd generate the right query and do exactly what you'd expect. Is there a way with the newer client to achieve the same result without introducing substantial additional complexity? Grasping at straws, I tried using SingleOrDefaultAsync(), but there isn't support for that, either. You can probably tell that most of my linq experience sources from working with EF, so I may just be trying to put a square peg in a round hole. 

Any pointers are appreciated.

**Comment by AlanWong-MS** [comment]
@mprigge, sorry for the late response, I was reallocated to another work item and had been out of office. Thanks for the feedback, we'll be adding back the Net45 library in our 7.4.0 beta 2 release (should be happening within this week).

**Comment by AlanWong-MS** [comment]
[OData Client 7.4.0 beta 2](https://www.nuget.org/packages/Microsoft.OData.Client/7.4.0-beta2) is out. The Nuget package now contains both the NetStandard 1.1 and Net45 versions.

**Comment by mprigge** [comment]
@AlanWong-MS Just so you're aware, the update to ODL 7.4.0 beta 2 is causing compatibility issues with the OData WebAPI libraries. Some really basic details here: https://github.com/OData/WebApi/issues/1193  This is maybe not an ODL problem per se, but it's notable that these issues don't appear in 7.4.0 beta, but do in beta 2.

**Comment by AlanWong-MS** [comment]
@mprigge thanks for the heads-up. I'll take a look at the thread you linked.

**Comment by victorperez2911** [comment]
@AlanWong-MS 
Is there any documentation for using the OData Client 7.4.0 beta 3 package ?

**Comment by AlanWong-MS** [comment]
@victorperez2911, we don't have specific documentation for the OData Client 7.4.0 beta 3 release. Fundamentally, the API is the same as it is with the 6.x release. Here's the [previous documentation](http://odata.github.io/odata.net/#04-01-basic-crud-operations). We're aiming to declare RTM very soon (likely early next week), and we may re-visit our documentation story if our resource allocation permits.

**Comment by philcarbone** [comment]
When I run the client generated from the Client Code Generator, I get the following runtime error:
`EdmParseException: Encountered the following errors when parsing the EDMX document: 
UnexpectedXmlAttribute : The attribute 'ConcurrencyMode' was not expected in the given context. : (8, 57)
UnexpectedXmlAttribute : The attribute 'ConcurrencyMode' was not expected in the given context. : (17, 57)
UnexpectedXmlAttribute : The attribute 'ConcurrencyMode' was not expected in the given context. : (37, 57)
UnexpectedXmlAttribute : The attribute 'ConcurrencyMode' was not expected in the given context. : (47, 57)
UnexpectedXmlAttribute : The attribute 'ConcurrencyMode' was not expected in the given context. : (66, 57)
UnexpectedXmlAttribute : The attribute 'ConcurrencyMode' was not expected in the given context. : (84, 57)
UnexpectedXmlAttribute : The attribute 'ConcurrencyMode' was not expected in the given context. : (101, 57)
UnexpectedXmlAttribute : The attribute 'ConcurrencyMode' was not expected in the given context. : (111, 57)
UnexpectedXmlAttribute : The attribute 'ConcurrencyMode' was not expected in the given context. : (124, 57)
UnexpectedXmlAttribute : The attribute 'Concurr...
`

which is throwing from this location:
`public static global::Microsoft.OData.Edm.IEdmModel GetInstance()
{
       return ParsedModel;
}`

Looks like it doesn't like "ConcurrencyMode"  :)

**Comment by AlanWong-MS** [comment]
@philcarbone, I've seen that problem before. It's been several weeks since I was able to work on the Client Code Gen, but I'm back on it for this week. Will have an update for it. Thanks for reporting!

**Comment by philcarbone** [comment]
Thanks! We've been upgrading to .NET Standard and .NET Core for some of our NuGet packages and using all this (now converging) OData Client code is really helping! We normally use the "OData Connected Service" plugin for visual studio 2017, but _for now_ we switched to the "OData v4 Client Code Generator v2.5.0", since there was a pre-release for it.

**Comment by tchuan** [comment]
@philcarbone there's a similar method which allows you to set a parameter to ignore UnsupportedAttribute.

**Comment by AlanWong-MS** [comment]
@philcarbone, looks the [removal of ConcurrencyMode](http://docs.oasis-open.org/odata/new-in-odata/v4.0/cn01/new-in-odata-v4.0-cn01.html#_Toc366145507
) is by design in OData V4.0 and removed in #564.

**Comment by philcarbone** [comment]
@AlanWong-MS and @tchuan -  Thanks for the info. So, how would I set a parameter to ignore the UnsupportedAttribute? The endpoint we are pulling from has ConcurrencyMode in the Metadata, so I need a way to work around that, if possible (other that a regex to remove the attributes from the generated code, which is what I'm doing now). Your help is greatly appreciated!

**Comment by philcarbone** [comment]
So I set `IgnoreUnexpectedElementsAndAttributes = true;` and it doesn't seem to work.

**Comment by AlanWong-MS** [comment]
@philcarbone, thanks for opening a new issue for the problem you found. I dug into it for a bit and I'll put my comments there.

**Comment by AlanWong-MS** [comment]
I've added the latest version of [OData Client Code Gen 2.5.0](http://vsixgallery.com/extension/ODataT4ItemTemplate.7a8976ef-0030-47ec-bd10-e8cdf7a2ccb2/) to VSIX Gallery server. To reiterate, the package itself is not signed as it's not an official package and the server is not an official feed supported by Microsoft.

Due to some change in VS2017, a few missing dependencies that I had to manually add after I used the code generator are:

1. System
2. System.Data
3. System.Data.DataSetExtensions
4. System.Net.Http
5. System.XML
6. System.Xml.Linq

Please feel free to check for issues on this pre-release package. We'll be releasing the official package on the Visual Studio Marketplace shortly.

**Comment by tchuan** [comment]
@AlanWong-MS I installed 2.5.0 and successfully generated the t4 template, but found some errors when Running Custom Tool. Severity	Code
`
Severity	Code	Description	Project	File	Line	Suppression State
Error		Running transformation: System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.OData.Edm, Version=7.5.0.20129, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
File name: 'Microsoft.OData.Edm, Version=7.5.0.20129, Culture=neutral, PublicKeyToken=31bf3856ad364e35'
`

**Comment by tchuan** [comment]
@AlanWong-MS  can not use this pre-release to generate client code. met this error:
`
Severity	Code	Description	Project	File	Line	Suppression State
Error		Running transformation: System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.OData.Edm, Version=7.5.0.20129, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
File name: 'Microsoft.OData.Edm, Version=7.5.0.20129, Culture=neutral, PublicKeyToken=31bf3856ad364e35'Severity	Code	Description	Project	File	Line	Suppression State
Error		There was a problem loading the assembly 'C:\USERS\TTAN\APPDATA\LOCAL\MICROSOFT\VISUALSTUDIO\15.0_F15D75A8\EXTENSIONS\OLOR0XRW.TCR\Microsoft.Spatial.dll' The following Exception was thrown:
System.IO.FileLoadException: Could not load file or assembly 'file:///C:\USERS\\APPDATA\LOCAL\MICROSOFT\VISUALSTUDIO\15.0_F15D75A8\EXTENSIONS\OLOR0XRW.TCR\Microsoft.Spatial.dll' or one of its dependencies. Strong name signature could not be verified.  The assembly may have been tampered with, or it was delay signed but not fully signed with the correct private key. (Exception from HRESULT: 0x80131045)1	

`

**Comment by robertmclaws** [comment]
@AlanWong-MS I've been noticing this for a while now, and ran into it myself in other ways. The change in assembly versioning strategy a few months ago is really going to screw everyone in non-obvious ways moving forward. Because the package version doesn't align with the assembly version, it is virtually impossible to determine what version should be specified in an AssemblyBindingRedirect without .NET Decompiler. You may think "well, just auto-generate AssemblyBindingRedirects" but you can't rely on that system, because it blows up constantly. I have a VS addin to fix binding redirects because it always generates duplicates for me.

You can have different AssemblyFileVersions if you want a way to keep track of build numbers, but please don't make end users deal with not having the build number match the package number... it's just not very nice :D

**Comment by AlanWong-MS** [comment]
@tchuan and @robertmclaws, thanks for the feedback! This was an issue I overlooked when I uploaded the unofficial VSIX package. We had thought about the possible binary and binary version mismatch situation between the released libraries themselves and the VSIX extension, and to shed some light, I've been working on our automation this past week to make sure the ODL binaries in the VSIX will be the same as the ones released.

I'm going to set up a MyGet with nightly releases of our VSIX package that aligns with the nightly releases of ODL.

**Comment by AlanWong-MS** [comment]
A few updates on the OData Client Code Gen VSIX package:
- I had intended to release a beta Client Code Gen along with the ODL 7.4.1 release from last week, but came across an issue while trying to post it onto the VS Marketplace; I've fixed the issue but it won't take place until the next release. I was also able to repro @tchuan's issue regarding the missing binaries and I've fixed that as well. Originally I was planning to upload the VSIX beta package here but due to the aforementioned missing binaries problem, I changed my mind as it's a huge issue that prevents people from using the extension. Thanks for pointing it out!
- [MyGet nightly Client Code Gen packages](https://www.myget.org/feed/odlnightly/package/vsix/ODataT4ItemTemplate.2592cbe8-b1fe-45b6-9a18-7518079af6a8) are now up (as of this morning) and the VSIX packages should have binaries that correspond with the Nightly packages from ODL side. The VSIX version scheme does not match exactly with ODL's due to VSIX versioning constraints, but you'll notice that the fourth VSIX version number (i.e. #.#.#.[this number]) should match the last eight values of the ODL nightly version (e.g. 7.5.0-Nightly2018**02221218**).

Known issue:
> You have to manually add the following dependencies:
> 1. System
> 2. System.Data
> 3. System.Data.DataSetExtensions
> 4. System.Net.Http
> 5. System.XML
> 6. System.Xml.Linq

- Speaking of VSIX version, we've decided to make it in the same format as ODL's to make it easier to know which ODL binaries are being installed. We understand that this change may cause some initial confusion; please let us know if there are any problems with this new version scheme.

Our next official ODL release _should_ have an official OData Client Code Gen release along with it.

**Comment by marvealmar** [comment]
Hi,
I am trying to use the code client code generator to create a client proxy. Cant get it to work, wonder if someone may help me out.

I have downloaded a ProjectOnline-emdx file. And downloaded the latest version of the code generator. 

NuGet packages:
![image](https://user-images.githubusercontent.com/368053/37027911-058d8bb2-2133-11e8-81cd-dced708fd8a0.png)

References:
![image](https://user-images.githubusercontent.com/368053/37028016-5cac55ae-2133-11e8-98dc-9df0914ebd9a.png)

Error messages:
![image](https://user-images.githubusercontent.com/368053/37028026-66d8452e-2133-11e8-8831-1d8b372c5895.png)

**Comment by AlanWong-MS** [comment]
@marvealmar, to clarify, which link did you download the latest version of Client Code Gen from?

**Comment by marvealmar** [comment]
@AlanWong-MS , I downloaded from: 
https://www.myget.org/feed/odlnightly/package/vsix/ODataT4ItemTemplate.2592cbe8-b1fe-45b6-9a18-7518079af6a8
I uninstalled previously installed version and retried as of today, installing version 7.5.0.03062136

**Comment by AlanWong-MS** [comment]
@marvealmar, thanks for the feedback. I downloaded the the nightly build and ran it against our [sample service](http://services.odata.org/V4/OData/OData.svc/), wrote a quick test for it, and it works for on my side. The error messages seem to imply that the dependencies (namely Microsoft.OData.Edm) aren't correctly added, but your screenshot shows that they've been added.

For reference, here's what my workspace (VS2017) looks like:
![image](https://user-images.githubusercontent.com/22382722/37113800-51746000-21fb-11e8-8eef-42627e69abbf.png)

![image](https://user-images.githubusercontent.com/22382722/37113835-67881b16-21fb-11e8-9a53-dd32c9666f57.png)

[OData v4 Client Code Generator 7.5.0.03062136](https://www.myget.org/feed/odlnightly/package/vsix/ODataT4ItemTemplate.2592cbe8-b1fe-45b6-9a18-7518079af6a8/7.5.0.03062136)
![image](https://user-images.githubusercontent.com/22382722/37113858-788d3356-21fb-11e8-8b56-3cdd583c8dbc.png)

Program.cs contents:
```
namespace ClientCodeGenTest
{
    using System;
    using System.Threading.Tasks;
    using Microsoft.OData.Client;

    class Program
    {
        static DemoService demoService = new DemoService(new Uri("http://services.odata.org/V4/OData/OData.svc/"));

        static void Main(string[] args)
        {
            var task = Task.Run(() => AsyncGetEntitySet());
            task.Wait();
        }

        static public async Task AsyncGetEntitySet()
        {
            var response = await demoService.Persons.ExecuteAsync();
            var people = response as QueryOperationResponse<Person>;

            foreach (var person in people)
            {
                Console.WriteLine(person.Name);
            }
        }
    }
}
```

Output:
![image](https://user-images.githubusercontent.com/22382722/37114032-032b03c6-21fc-11e8-9ba0-62cb46ee7e05.png)

If you don't see anything that stands out to you, would you be able to share the content of your EDMX file?

**Comment by marvealmar** [comment]
Hi, @AlanWong-MS 
I tried to connect to your sample service as well, using
public const string MetadataDocumentUri = "http://services.odata.org/V4/Odata/OData.svc/";
I have the same packages included and referenced dll's as in  your sample.

Still have the same problem, so I guess it is not related to my edmx-file?

**Comment by cilerler** [comment]
@marvealmar are you sure you did replace the `public const string MetadataDocumentUri = "http://services.odata.org/V4/Odata/OData.svc/";` part with your own address?

**Comment by marvealmar** [comment]
@cilerler , yes I am. I also tries with the http://services.odata.org/V4/Odata/OData.svc/" as this seems to be a public testable service. It returns a simple Json at least.

What I am trying to achieve is to connect to a ProjectOnline OData service. As this has some credentials constraints, I have to download and save the file as a local edmx-file. So I have in fact been trying type:
MetadataDocumentUri = @"C:\ProjectOnline.edmx";

But as it neither seems to work with my local file nor the testable service, I am quite stumbled:)

**Comment by AlanWong-MS** [comment]
@marvealmar, yes, those are the things I would have suggested to work around credential-based services.

Hm, I actually didn't think it's related to the EDMX file based on the error messages, but I was wondering about the contents in case there was something I was missing. It looked more like an issue with one or more of the OData libs being referenced.

Would it be possible for you to upload your solution and project (including the OData lib references) to a repro? I can sync and take a look. Not too sure what the issue could be at the moment.

**Comment by marvealmar** [comment]
@AlanWong-MS 
I dont have a repro. Is it ok to just upload her as a zip-file?
[ConsoleApp3.zip](https://github.com/OData/odata.net/files/1802839/ConsoleApp3.zip)

If not, I will create a repro

**Comment by AlanWong-MS** [comment]
@marvealmar, the zip is fine, thanks. I downloaded it and ran the latest nightly Code Gen on the project and it generated the proxy .cs file fine on my side (against the http://services.odata.org/V4/OData/OData.svc/ service). I also ran file diffs between our .tt and .ttinclude files and they're identical.

Here are the exact steps that I took to get my Code Gen set up:
1. Download and install latest [Client Code Gen](https://www.myget.org/feed/odlnightly/package/vsix/ODataT4ItemTemplate.2592cbe8-b1fe-45b6-9a18-7518079af6a8).
2. Create a new console app project in Visual Studio 2017 (Community Edition works fine for me).
3. Right-click on the console app project in the Solution Explorer and go to "Add->New Item...".
4. Select OData Client and click "Add" (default settings should be fine). At this point, the Client Code Gen extension will automatically add the OData library, packages.config, ODataClient1.tt, and ODataClient1.ttinclude.
5. In the Solution Explorer, double-click the ODataClient1.tt that was just added to open the file.
6. In the file, add "http://services.odata.org/V4/OData/OData.svc/" to the MetadataDocumentUri field.
7. A new ODataClient1.cs file should be generated.
![image](https://user-images.githubusercontent.com/22382722/37303606-31160508-25ec-11e8-9e4c-734240362d0b.png)

If those are more or less the same steps that you took and you still experience the issue, then the problem could be caused by a difference in our Visual Studio environments.

**Comment by KeffiH** [comment]
I have tried reproducing the sample project and in another project. The .cs file is fine but when I run the project it fails on line 31 - this.Format.UseJson();. Seems the same on both projects I have tried.

**Comment by AmitBhatnagar24** [comment]
Come on. This is crazy. Over 2 days spent chasing rabbit holes in trying to get this and the library to work. Following the steps @AlanWong-MS outlined above, all I get is:

`Compiling transformation: The type or namespace name 'CsdlReaderSettings' could not be found (are you missing a using directive or an assembly reference?)`

How are we supposed to get anything to work with the current state of OData.NET ?? Is the recommended approach to fallback to v6?!?!?

**Comment by AlanWong-MS** [comment]
@KeffiH, was it a compilation error regarding not being able to find that method?
@AmitBhatnagar24, thanks for the feedback. Looks like at least two people are experiencing that issue then. I don't have an answer as to why that's happening; I've tried the process on two separate machines (one of which is pretty clean VM) and they both work for me. cc: @marvealmar

Here's an attachment of the small test project I wrote. It compiles and runs successfully on my machine as seen in the screenshots above. If even that doesn't work, I think maybe there's some sort of conflicting setting somewhere...

[ClientCodeGenTest.zip](https://github.com/OData/odata.net/files/1834256/ClientCodeGenTest.zip)

**Comment by henning-krause** [comment]
I just upgraded to v7.x and experienced the exactly the same issue as @AmitBhatnagar24.

Turned out, I had two competing extensions installed: The Connected Service Client Generator (Odata 1-4) as well as the shiny new Odata Client Generator v7.x.

I uninstalled both and reinstalled the Odata Client Generator and the problem went away.

**Comment by AlanWong-MS** [comment]
@henning-krause, thanks for the response! I was in the middle of looking through my `C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2017\VERSION\COMMON7\IDE\EXTENSIONS\` and was just wondering about the same thing. Thanks for pointing out the fix!

**Comment by mprigge** [comment]
@AlanWong-MS I've observed the same problem - some members of my team have run into that. This may be a total newb question as I have never written/maintained a T4-based extension, but is there a reason that the code generation script couldn't reference/use the ODL that's referenced by the project it lives in? Generally speaking, you're never going to see that extension utilized in a project that doesn't have the ODL references already (the output code would be useless if you did). Not even sure if that's even a thing that could be done, but you'd neatly avoid this kind of complexity if it were possible. Completely disregard if not.

**Comment by KeffiH** [comment]
The zip you uploaded ClientCodeGenTest works OK first time. The other version I made on my machine from your previous sample, which seems identical, has this problem with the Format JSON. Both run on same machine and VS 2017 Community. Strange - all solutions I have tried build without error. Will search further.

**Comment by AmitBhatnagar24** [comment]
confirmed that the work around that @henning-krause found gets me past the errors I reported. I too had both extensions installed and removing them both, and then reinstalling the latest beta version (linked above) got me past the running of the tt.

However, it seems that the tool can't work with an older OData version? This is what I see when I obtain the metadata for the OData service I need to hit (with authentication)

```
<?xml version="1.0" encoding="utf-8"?>
<edmx:Edmx Version="1.0" xmlns:edmx="http://schemas.microsoft.com/ado/2007/06/edmx">
```

I believe this is saying OData v 1.0? I get the same result even if I try to pass in a `OData-Version` header with a value of 4.0.

What are my options now?

**Comment by AmitBhatnagar24** [comment]
(is there a gitter or a slack channel we can all go to continue the conversation?)

**Comment by KeffiH** [comment]
@AlanWong-MS  Established now that it is a versioning problem with the code gererator/client. Your zipped solution worked fine, but as soon as I tried to run the tool again on my machine  I had the same problems. So its a question of which works with what at the moment.

**Comment by marvealmar** [comment]
Hi @AlanWong-MS (and others),
Sorry for not being very responsive on my reported error. I had to attend some other stuff. I see that my issue most likely is caused by whats reported by @henning-krause. I had a OData Connected Service tool installed, but wanted to try to replace this with the Client Generator Tool. So I have these two installations. I also tried several versions of the Client Generator by the way. 
I will not be able to look into this again for some time, so I assume that the proposed fix as described by @henning-krause will work. 
Thanks for all help!!

**Comment by AlanWong-MS** [comment]
@AmitBhatnagar24, the extension doesn't support authenticated endpoints at this time. A workaround is to download the metadata as a text file and point the MetadataDocumentUri field to your local copy of the metadata.

@KeffiH, ideally try to use the same ODL version packaged in the extension. Otherwise, the compiler will complain about different versions of ODL being used in the project.

When you say that running the tool again on the machine and you came across the problem, were you running the tool against my solution a second time and it failed? Or was it a different project? I think that this may be caused by a different ODL version cached somewhere again and Visual Studio is trying to read from the old version.

@marvealmar, no problem, just wanted to tag you so that you're aware of the conversation.

**Comment by AmitBhatnagar24** [comment]
Can someone please point me to steps on how I can consume an OData v3 service? I can't seem to find the right client / code generator / service extension combination that will allow me to continue anything now.

**Comment by AlanWong-MS** [comment]
@AmitBhatnagar24, based on the [version history](https://marketplace.visualstudio.com/items?itemName=bingl.ODatav4ClientCodeGenerator), unfortunately, I don't think there were any released Client Code Generators that support a v3 service. For versions that support v4 service and ODL 6.x, here is a [repo](https://github.com/OData/lab/tree/Tools/Tools) with previous releases.

**Comment by KeffiH** [comment]
@AlanWong-MS I removed your tt and replaced it with one generated from my setup. The error message I am getting seems to suggest it is having a problem with "ConcurrencyMode". Which is odd as I am just using the sample endpoint.

**Comment by KeffiH** [comment]
@AlanWong-MS Just to update. I set everything to 7.4.3 from NuGet and although code generated I still had a problem with running. Removed a "rownumber" column (I use entity Framework) from the tables and all now runs OK.

**Comment by AlanWong-MS** [comment]
@KeffiH, The ConcurrencyMode problem is a known issue. I forgot to mention that; this was due to a change between ODL 6.x and 7.x that deprecated the "ConcurrencyMode" attribute. #1044 is opened against the bug and related issues. A quick workaround is just to search for the "ConcurrencyMode" in the model and remove the line.

**Comment by KeffiH** [comment]
Yes, work around fixed it.

**Comment by AlanWong-MS** [comment]
Closing this thread as we've released ODL Client 7.x. I have a separate PR #1150 to address some Code Gen bugs. Should anyone find any new bugs related to Client, please open new issues. Thanks everyone for being supportive throughout this journey!

**Comment by victorperez2911** [comment]
Hi @AlanWong-MS 

How do I use open types with this library?
